src/library/scala/collection/immutable/TreeMap.scala

```diff
@@ -29,34 +33,23 @@ object TreeMap extends ImmutableSortedMapFactory[TreeMap] {
   implicit def canBuildFrom[A, B](implicit ord: Ordering[A]): CanBuildFrom[Coll, (A, B), TreeMap[A, B]] = new SortedMapCanBuildFrom[A, B]
 
   override def newBuilder[A, B](implicit ord: Ordering[A]): mutable.Builder[(A, B), TreeMap[A, B]] = new TreeMapBuilder
-  private final class TreeMapBuilder[A, B](implicit ordering: Ordering[A]) extends mutable.Builder[(A, B), TreeMap[A, B]] {
+
+  private class TreeMapBuilder[A, B](implicit ordering: Ordering[A])
+    extends RB.MapHelper[A, B]
+      with Builder[(A, B), TreeMap[A, B]] {
     type Tree = RB.Tree[A, B]
-    private [this] var tree:Tree = null
-    override def +=(elem: (A, B)): this.type = {
-      tree = RB.update(tree, elem._1, elem._2, overwrite = true)
+    private var tree:Tree = null
+
+    def +=(elem: (A, B)): this.type = {
+      tree = mutableUpd(tree, elem._1, elem._2)
       this
     }
-
-    private object adder extends Function2[A, B, Unit] {
-      var accumulator :Tree = null
-      def addTree(aTree:Tree, bTree: Tree): Tree = {
-        // TODO should consider non overlapping trees
-        // just bulk add the non overlapping parts and
-        // only addAll for the intersecting range
-        //
-        // TODO consider adding the smaller map to the larger one with a flag to
-        // say who wins
-
-        accumulator = aTree
-        RB.foreachEntry(bTree, this)
-
-        val result = accumulator
-        // be friendly to GC
-        accumulator = null
-        result
-      }
+    private object adder extends AbstractFunction2[A, B, Unit] {
+      // we cache tree to avoid the outer access to tree
+      // in the hot path (apply)
+      private[this] var accumulator :Tree = null
       def addForEach(aTree:Tree, hasForEach: HasForeachEntry[A, B]): Tree = {
```

Comment:
The parameter `aTree` is unused here.

Reply:
removed and reworked `addForEach` to be mutating the tree, rather than returning it